### DISCUSSION TITLE
What is the method to Fill empty spaces in a Path?

### DISCUSSION THREAD

**Discussion by rafasashi** [fabricjs/fabric.js#10231] [Q&A]
Is there a method to fill the empty space in the middle of a path with a solid background color.

My goal is to trace a background fill that covers the entire shape, including filling in the gaps.

![c17a1d69d7104d460766a256574b207d 1](https://github.com/user-attachments/assets/0ddf5e65-3224-4d2e-b09a-bae0594bcd1a)

I have tried to truncate path after second 'M' command based on the following answer:

https://graphicdesign.stackexchange.com/questions/166236/what-is-the-method-to-fill-gaps-in-an-svg-path-with-color

However if the shape is composed of separated elements then only one element is drawn.

![Screenshot (933)](https://github.com/user-attachments/assets/523d2d3e-ed8f-4484-b71b-8d7185a56b65)

So instead I tried to split all subpath and add them as a unique object. The overlapping gave the result that I was expecting in most cases:

![Screenshot (934)](https://github.com/user-attachments/assets/c432f183-7082-4e5c-84a4-a17d64ed8dce)

However the performance is very bad for larger objects and I can't implement it that way.

So now I am trying to find a way to detect empty shapes  and either draw an inverted path on top or discard them to speedup the splitting process.

Alternatively detecting the largest object and/or disconnected large objects is something that I have tried to do with boolean operations via paper.js but exporting and importing path is not giving good results. 

Any suggestions would be greatly appreciated!

**Comment by rafasashi** [comment]
I also tried inverting and clipping the shape on top of a colored rectangle with different combinations of globalCompositeOperation but I always end-up with a colored surrounding which I can't exclude:

![Screenshot (932)](https://github.com/user-attachments/assets/9441d151-a8e8-412c-b228-cf815f093a07)

**Comment by asturur** [comment]
I do not think you can find a way that covers all the path use cases.
You have to try to change the fillRule property between the 2 available values and see if it gives you some coverage.

Who controls the paths?
If they are from a static library you should just provide a path for each path that can accomplish this task.

You can't automate the process without using some  tool that has path knowledge like paper.js indeed.

**Comment by rafasashi** [comment]
Thanks, @asturur! For now, I'll be using a default library of objects and manually "preparing" the background path as needed.

In the future, I'm hoping to allow users to import objects directly, ideally by automating processes like union, simplification, and inversion. 

However, I think the first step will be finding a way to go from rasterizing to tracing (possibly with AI), optimizing, and then drawing.

**Reply by asturur** [reply]
but also try fillRule. I don't see it mentioned in your examples. It should do the same thing that you separating the objects in many paths is doing, but with no performance cost.

**Reply by rafasashi** [reply]
Perhaps I am doing something wrong, but neither nonzero nor evenodd seem to make any difference. I've tried using different paths, but nothing seems to change.

 ```

	var attr = {
		
		id: group.id + '_background',
		fillRule: 'nonzero', //<---------------- tried here
					
	};

	self.getGroupAttrs().forEach(function(i) {
		
		if( typeof attr[i] === 'undefined' && typeof group[i] !== 'undefined' ) {
			
			attr[i] = group[i];
		}
	});

	var clonedGroup = new fabric.Group([],attr);
	
	group.getObjects().forEach(function(obj,index) {

		if (obj.type === 'path') {
			
			obj.clone(function(clonedObj) {
						
				clonedObj.set({
						
					id: group.id + '_path_' + index,
					fill: 'red',
					fillRule: 'nonzero', //<---------------- and tried here
					strokeWidth: 0,
				});
					
				clonedGroup.add(clonedObj);
			});
		}
	});
```

**Reply by rafasashi** [reply]
However, it works with the polygon from the following example: https://www.geeksforgeeks.org/fabric-js-polygon-fillrule-property/

**Reply by rafasashi** [reply]
I am using 5.3.1 maybe it's a problem...

**Reply by asturur** [reply]
no is not a problem per se.
Make a test with caching off in case fillRule is strangely out of the cache invalidating properties.
If it doesn't give you any difference it can be just that the paths are identical under the 2 conditions. Not all the paths react the same.
That is why i told you, you can't make this with a blanket approach
